### PR TITLE
[AIR] Support array output from Torch model

### PR DESCRIPTION
**Pull request by amogkam**
The previous implementation of `TorchPredictor` failed when the model outputted an array for each sample (for example outputting logits). This is because Pandas Dataframes cannot hold numpy arrays of more than 1 element. We have to convert the outermost numpy array returned by the model to a list so that it be can stored in a Pandas Dataframe.

The newly added test fails under the old implementation.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 1]
```diff
@@ -82,7 +82,11 @@ def _convert_to_tensor(
     def _predict(self, tensor: torch.Tensor) -> pd.DataFrame:
         """Handle actual prediction."""
         prediction = self.model(tensor).cpu().detach().numpy()
-        return pd.DataFrame(prediction, columns=["predictions"])
+        # If model has outputs a Numpy array (for example outputting logits),
+        # these cannot be used as values in a Pandas Dataframe.
+        # We have to convert the outermost dimension to a python list (but the values
+        # in the list can still be Numpy arrays).
+        return pd.DataFrame({"predictions": list(prediction)}, columns=["predictions"])
```
Comment (bveeramani) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:89]: Can we use `TensorArray` here now that #24752 is merged?

```suggestion
        return pd.DataFrame({"predictions": TensorArray(prediction)}, columns=["predictions"])
```

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 2]
```diff
@@ -112,10 +116,10 @@ def predict(
...
                 case it will be applied to all tensors.
                 If None, then automatically infer the dtype.
-            unsqueeze_feature_tensors (bool): If set to True, the features tensors
-                will be unsqueezed (reshaped to (N, 1)) before being concatenated into
-                the final features tensor. Otherwise, they will be left as is, that is
-                (N, ). Defaults to True.
+            unsqueeze (bool): If set to True, the features tensors will be unsqueezed
+                (reshaped to (N, 1)) before being concatenated into the final features
+                tensor. Otherwise, they will be left as is, that is (N, ).
+                Defaults to True.
```
Comment (bveeramani) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:122]: Nit: out of scope for this PR, but this explanation doesn't really make sense for multi-dimensional inputs

File: python/ray/ml/tests/test_torch_predictor.py [hunk 3]
```diff
@@ -45,6 +45,20 @@ def test_predict():
     assert predictions.to_numpy().flatten().tolist() == [4, 8, 12]
 
 
+def test_predict_array_output():
+    """Tests if predictor works if model outputs an array instead of single value."""
+
+    predictor = TorchPredictor(model=model, preprocessor=preprocessor)
```
Comment (bveeramani) [inline python/ray/ml/tests/test_torch_predictor.py:51]: Nit: this is definitely out of scope for this PR, but it might be better to define `model` and `preprocessor` in the test because:
* Less confusion when reading the tests since all of the code is visible from within each test.
* Less chance of sharing state between tests, which creates unwanted dependencies between them.

[Related advice](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#why-5) from .NET's unit testing best practices.

Comment (bveeramani) [inline python/ray/ml/tests/test_torch_predictor.py:51]: Just for my own understanding, why do we need to use a preprocessor here?

**Review by bveeramani — COMMENTED** [review]
Overall LGTM. If we can't use `TensorArray` here, then I think this should be good to merge

File: python/ray/ml/tests/test_torch_predictor.py (hunk 3 above)
Comment (Yard1) [reply python/ray/ml/tests/test_torch_predictor.py:51]: I believe it is to check that its applied, though that could be covered by other tests and not this one specifically

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 2 above)
Comment (Yard1) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:122]: We could change the description, yeah. We are now calling `unsqueeze` which should work with an arbitrary number of dimensions and not just 2, like the `view` call we were using before.

**Review by Yard1 — APPROVED** [review]
Thanks, this looks good. +1 to Balaji's comment about using a `TensorArray` once that's possible, though we probably don't need to block on that.

File: python/ray/ml/tests/test_torch_predictor.py (hunk 3 above)
Comment (krfricke) [reply python/ray/ml/tests/test_torch_predictor.py:51]: In this case I'd vote to leave it out of this test. We should ideally only test one thing

**Review by krfricke — APPROVED** [review]
Generally looks good once @bveeramani's comments are addressed

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 2 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:122]: Agreed the semantics here need to be changed. Though this is out of scope for this PR- this PR is just updating the docstring to match the same as the argument name.

File: python/ray/ml/tests/test_torch_predictor.py (hunk 3 above)
Comment (amogkam) [reply python/ray/ml/tests/test_torch_predictor.py]: Sounds good,  I updated the tests!

Comment (amogkam) [reply python/ray/ml/tests/test_torch_predictor.py]: Updated!

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 1 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:89]: I didn't want to have a ray.data dependency in Predictors. Once we do the data type conversion cleanup and probably move `TensorArray` to a common utils file, I definitely agree though, we should use TensorArray here!

**Review by bveeramani — APPROVED** [review]
LGTM
